### ISSUE TITLE
Cannot switch variants on macOS

### ISSUE THREAD

**Issue by leroyvn** [mitsuba-renderer/mitsuba2#56]
Using the latest master, switching variants has become impossible.

# System configuration

Platform: macOS 10.15.4
Compiler: Apple clang version 11.0.0 (clang-1100.0.33.17)
Python version: Python 3.7.5 :: Anaconda, Inc.
Mitsuba 2 version: e8b5971e6942cc5cb92d7d088535f52a66e4d070 with patch to compile with packet variant (leroyvn@a1aac2f61265f834d23c9cb950f4a667f9f85030)
Compiled variants: "scalar_mono", "scalar_mono_polarized", "scalar_rgb", "scalar_spectral", "packet_rgb"

# Problem

I cannot switch variants anymore without unloading Mitsuba first.

# Steps to reproduce

1. Build the code with the aforementioned settings.
2. Run pytest. At this stage, a suspiciously large number of tests is skipped.
   <details>
   <summary>Click to see pytest log</summary>
   
   ```
      Running the full test suite. To skip slow tests, please run 'pytest -m 'not slow'
   ============================= test session starts ==============================
   platform darwin -- Python 3.7.7, pytest-5.4.1, py-1.8.1, pluggy-0.13.1
   rootdir: /Users/vincent/Documents/src/thirdparty/mitsuba2, inifile: setup.cfg
   plugins: xdist-1.30.0, forked-1.1.3
   collected 989 items
   
   src/bsdfs/tests/test_blendbsdf.py .....                                  [  0%]
   src/bsdfs/tests/test_conductor.py sss                                    [  0%]
   src/bsdfs/tests/test_dielectric.py .....                                 [  1%]
   src/bsdfs/tests/test_diffuse.py ..s                                      [  1%]
   src/bsdfs/tests/test_polarizer.py sssss                                  [  2%]
   src/bsdfs/tests/test_retarder.py sssssss                                 [  2%]
   src/bsdfs/tests/test_rough_conductor.py sssss                            [  3%]
   src/bsdfs/tests/test_rough_dielectric.py sssssssssss                     [  4%]
   src/bsdfs/tests/test_rough_plastic.py ss                                 [  4%]
   src/bsdfs/tests/test_twosided.py ...                                     [  4%]
   src/emitters/tests/test_area.py .ssssss                                  [  5%]
   src/emitters/tests/test_constant.py sssss                                [  6%]
   src/emitters/tests/test_point.py ssssss                                  [  6%]
   src/films/tests/test_hdrfilm.py .....                                    [  7%]
   src/libcore/tests/test_argparser.py .....                                [  7%]
   src/libcore/tests/test_atomic.py .                                       [  7%]
   src/libcore/tests/test_bbox.py ...                                       [  8%]
   src/libcore/tests/test_bitmap.py ............                            [  9%]
   src/libcore/tests/test_bsphere.py .                                      [  9%]
   src/libcore/tests/test_distr_1d.py ssssssssssssssssss                    [ 11%]
   src/libcore/tests/test_distr_2d.py ..............................sssssss [ 15%]
   ssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssss [ 22%]
   sssssssssssssssssssssssssssssssssssssssss                                [ 26%]
   src/libcore/tests/test_filesystem.py .............                       [ 27%]
   src/libcore/tests/test_frame.py ...                                      [ 28%]
   src/libcore/tests/test_logger.py .                                       [ 28%]
   src/libcore/tests/test_math.py ..s.....                                  [ 29%]
   src/libcore/tests/test_mmap.py ....                                      [ 29%]
   src/libcore/tests/test_properties.py .......s                            [ 30%]
   src/libcore/tests/test_qmc.py ...s                                       [ 30%]
   src/libcore/tests/test_quad.py ....                                      [ 31%]
   src/libcore/tests/test_random.py ..s                                     [ 31%]
   src/libcore/tests/test_spline.py ....s.s......                           [ 32%]
   src/libcore/tests/test_stream.py ............................            [ 35%]
   src/libcore/tests/test_struct.py ....................................... [ 39%]
   ........................................................................ [ 46%]
   ........................................................................ [ 53%]
   .............                                                            [ 55%]
   src/libcore/tests/test_transform.py ......                               [ 55%]
   src/libcore/tests/test_util.py ..                                        [ 56%]
   src/libcore/tests/test_vector.py .                                       [ 56%]
   src/libcore/tests/test_warp.py sss..sssssss....                          [ 57%]
   src/libcore/tests/test_xml.py .....................                      [ 59%]
   src/librender/tests/test_bsdf.py ..                                      [ 60%]
   src/librender/tests/test_fresnel.py ..sss..                              [ 60%]
   src/librender/tests/test_imageblock.py ...sss                            [ 61%]
   src/librender/tests/test_integrator.py ...ssssss...ssssss...ssssss...sss [ 64%]
   sss...ssssss...sss                                                       [ 66%]
   src/librender/tests/test_interaction.py ..s                              [ 66%]
   src/librender/tests/test_kdtrees.py ..s                                  [ 67%]
   src/librender/tests/test_mesh.py .......................                 [ 69%]
   src/librender/tests/test_microfacet.py .sssssssssssssssssssssssssssss    [ 72%]
   src/librender/tests/test_mueller.py ........                             [ 73%]
   src/librender/tests/test_records.py .ss.s                                [ 73%]
   src/librender/tests/test_renders.py ssssssssssssssssssssssssssssssssssss [ 77%]
   ..................ssssssssssssssssssssssssssssssssssss                   [ 83%]
   src/librender/tests/test_scene.py .                                      [ 83%]
   src/librender/tests/test_spectra.py .sssss                               [ 83%]
   src/librender/tests/test_spiral.py ...                                   [ 84%]
   src/phase/tests/test_hg.py .s                                            [ 84%]
   src/phase/tests/test_isotropic.py ..s                                    [ 84%]
   src/phase/tests/test_trampoline.py ..                                    [ 84%]
   src/rfilters/tests/test_rfilter.py .........                             [ 85%]
   src/samplers/tests/test_independent.py ...s                              [ 86%]
   src/sensors/tests/test_perspective.py ................ssssssssssss       [ 88%]
   src/sensors/tests/test_thinlens.py ................sssssssssssssssssssss [ 92%]
   sssssssssssssssssssssssssssssssssssssssssssssssssssssss                  [ 98%]
   src/shapes/tests/test_cylinder.py ...                                    [ 98%]
   src/shapes/tests/test_disk.py ...                                        [ 98%]
   src/shapes/tests/test_rectangle.py ..s                                   [ 99%]
   src/shapes/tests/test_sphere.py ....                                     [ 99%]
   src/spectra/tests/test_d65.py s                                          [ 99%]
   src/spectra/tests/test_irregular.py ss                                   [ 99%]
   src/spectra/tests/test_regular.py ss                                     [100%]
   
   =========================== short test summary info ============================
   SKIPPED [16] /Users/vincent/Documents/src/thirdparty/mitsuba2/src/conftest.py:42: Mitsuba variant "scalar_mono_polarized" is not enabled!
   SKIPPED [198] /Users/vincent/Documents/src/thirdparty/mitsuba2/src/conftest.py:42: Mitsuba variant "packet_rgb" is not enabled!
   SKIPPED [104] /Users/vincent/Documents/src/thirdparty/mitsuba2/src/conftest.py:42: Mitsuba variant "packet_spectral" is not enabled!
   SKIPPED [12] /Users/vincent/Documents/src/thirdparty/mitsuba2/src/conftest.py:42: Mitsuba variant "scalar_spectral" is not enabled!
   SKIPPED [1] /Users/vincent/Documents/src/thirdparty/mitsuba2/src/libcore/tests/test_math.py:41: packet_rgb mode not enabled
   SKIPPED [1] src/libcore/tests/test_properties.py:149: TODO fix AnimatedTransform
   SKIPPED [1] src/libcore/tests/test_qmc.py:100: RadicalInverse has no vectorized bindings
   SKIPPED [10] /Users/vincent/Documents/src/thirdparty/mitsuba2/src/libcore/tests/test_warp.py:23: packet_rgb mode not enabled
   SKIPPED [1] /Users/vincent/Documents/src/thirdparty/mitsuba2/src/librender/tests/test_imageblock.py:111: packet_rgb mode not enabled
   SKIPPED [1] /Users/vincent/Documents/src/thirdparty/mitsuba2/src/librender/tests/test_imageblock.py:160: packet_rgb mode not enabled
   SKIPPED [15] /Users/vincent/Documents/src/thirdparty/mitsuba2/src/conftest.py:79: Mitsuba variant "packet_rgb" is not enabled!
   SKIPPED [15] /Users/vincent/Documents/src/thirdparty/mitsuba2/src/conftest.py:79: Mitsuba variant "gpu_rgb" is not enabled!
   SKIPPED [3] /Users/vincent/Documents/src/thirdparty/mitsuba2/src/conftest.py:69: Mitsuba variant "packet_rgb" is not enabled!
   SKIPPED [18] /Users/vincent/Documents/src/thirdparty/mitsuba2/src/librender/tests/test_renders.py:17: scalar_mono mode not enabled
   SKIPPED [18] /Users/vincent/Documents/src/thirdparty/mitsuba2/src/librender/tests/test_renders.py:17: scalar_mono_polarized mode not enabled
   SKIPPED [18] /Users/vincent/Documents/src/thirdparty/mitsuba2/src/librender/tests/test_renders.py:17: scalar_spectral mode not enabled
   SKIPPED [18] /Users/vincent/Documents/src/thirdparty/mitsuba2/src/librender/tests/test_renders.py:17: packet_rgb mode not enabled
   SKIPPED [1] /Users/vincent/Documents/src/thirdparty/mitsuba2/src/samplers/tests/test_independent.py:53: packet_rgb mode not enabled
   SKIPPED [1] /Users/vincent/Documents/src/thirdparty/mitsuba2/src/shapes/tests/test_rectangle.py:89: packet_rgb mode not enabled
   ================= 537 passed, 452 skipped in 115.41s (0:01:55) =================
   ```
   </details>
3. Import test: one variant at a time is fine.
    ```
    $ python -c "import mitsuba; mitsuba.set_variant('scalar_rgb')"
    $ python -c "import mitsuba; mitsuba.set_variant('packet_rgb')"
    ```
4. Import test: load a variant then switch raises.
   ```
   $ python -c "import mitsuba; mitsuba.set_variant('scalar_rgb'); mitsuba.set_variant('packet_rgb')"
   Traceback (most recent call last):
     File "<string>", line 1, in <module>
     File "/Users/vincent/Documents/src/thirdparty/mitsuba2/build/dist/python/mitsuba/__init__.py", line 162, in set_variant
       _import('mitsuba.core_' + value + '_ext')),
     File "/Users/vincent/miniconda3/envs/mitsuba2/lib/python3.7/importlib/__init__.py", line 127, in import_module
       return _bootstrap._gcd_import(name[level:], package, level)
     File "<frozen importlib._bootstrap>", line 1006, in _gcd_import
     File "<frozen importlib._bootstrap>", line 983, in _find_and_load
     File "<frozen importlib._bootstrap>", line 967, in _find_and_load_unlocked
     File "<frozen importlib._bootstrap>", line 670, in _load_unlocked
     File "<frozen importlib._bootstrap>", line 583, in module_from_spec
     File "<frozen importlib._bootstrap_external>", line 1043, in create_module
     File "<frozen importlib._bootstrap>", line 219, in _call_with_frames_removed
   ImportError: generic_type: type "Properties" is already registered!
   ```

Could this have something to do with my build setup?

**Comment by abhinavvs** [comment]
I have the same issue on Linux (Ubuntu 16.04) as well. I had to **restart the Python Kernel** and only then was I able to switch variants. 

Here is the error message I get when switching from `scalar_rgb` to `gpu_autodiff_rgb` (which is almost exactly same as the one shown above):

```python
ImportError                               Traceback (most recent call last)
<ipython-input-46-d33d0fecbe85> in <module>
----> 1 mitsuba.set_variant('gpu_autodiff_rgb')

~/mitsuba2/build/dist/python/mitsuba/__init__.py in set_variant(value)
    160         modules = {
    161             'mitsuba.core': (_import('mitsuba.core_ext'),
--> 162                              _import('mitsuba.core_' + value + '_ext')),
    163             'mitsuba.render': (_import('mitsuba.render_ext'),
    164                                _import('mitsuba.render_' + value + '_ext'))

~/anaconda3/lib/python3.7/importlib/__init__.py in import_module(name, package)
    125                 break
    126             level += 1
--> 127     return _bootstrap._gcd_import(name[level:], package, level)
    128 
    129 
~/anaconda3/lib/python3.7/importlib/_bootstrap.py in _gcd_import(name, package, level)
~/anaconda3/lib/python3.7/importlib/_bootstrap.py in _find_and_load(name, import_)
~/anaconda3/lib/python3.7/importlib/_bootstrap.py in _find_and_load_unlocked(name, import_)
~/anaconda3/lib/python3.7/importlib/_bootstrap.py in _load_unlocked(spec)
~/anaconda3/lib/python3.7/importlib/_bootstrap.py in module_from_spec(spec)
~/anaconda3/lib/python3.7/importlib/_bootstrap_external.py in create_module(self, spec)
~/anaconda3/lib/python3.7/importlib/_bootstrap.py in _call_with_frames_removed(f, *args, **kwds)

ImportError: generic_type: type "Ray3f" is already registered!
```

**Comment by wjakob** [comment]
cc @Speierers

**Comment by Speierers** [comment]
My bad! This is now fixed and I added a test for this: b7b7f1d
Thanks for reporting this!
